### ISSUE TITLE
Loss Instability

### ISSUE THREAD

**Issue by andyhandsom6** [physical-superintelligence-lab/Psi0#43]
Hi, great work! I have a quick question about loss calculation. 

I've noticed that the loss isn't averaged on the chunk length dimension. That might lead to a problem: As the prefix of training-time RTC is masked and doesn't count into loss, and the length of prefix is randomized across samples, this could mean that the action terms (in a chunk) counting into loss varies across samples, which would affect the scale of loss. Would this affect training stability? 

Looking forward to hearing your thoughts on this!

**Comment by songlin** [comment]
Hi @andyhandsom6, 

Good observation! As you noticed, we apply the actions_mask and prefix_mask before computing the loss, as shown below. Specifically, actions_mask represents the masks generated during the action transforms, such as timestep padding or action-dimension padding. And prefix_mask corresponds to the training-time RTC token masks. We first perform a SUM over the chunk dimension to keep the gradient magnitude consistent across different batches, and then perform a MEAN over the batch dimension. In practice, I never observed loss instability during the fine-tuning stage.

https://github.com/physical-superintelligence-lab/Psi0/blob/e100324f76349619599e0ca21567577c4b0f8b26/src/psi/trainers/finetune.py#L586-L598

Best,
Songlin

**Comment by andyhandsom6** [comment]
I see, the expectation of # of terms counting into loss should be consistent across batches. Thanks for your quick response!
